File: mobile/apps/photos/plugins/ente_background_manager/android/src/main/kotlin/io/ente/background/TaskConfiguration.kt
package io.ente.background

import org.json.JSONObject

internal data class TaskConfiguration(
    val identifier: String,
    val kind: String,
    val frequencyMs: Long,
    val initialDelayMs: Long,
    val flexMs: Long?,
    val requiresNetwork: Boolean,
    val requiresCharging: Boolean,
    val requiresDeviceIdle: Boolean,
    val runBudgetMs: Long?,
    val foregroundStopTimeoutMs: Long?,
    val callbackHandle: Long,
) {
    fun encode(): String =
        JSONObject(
                mapOf(
                    "identifier" to identifier,
                    "kind" to kind,
                    "frequencyMs" to frequencyMs,
                    "initialDelayMs" to initialDelayMs,
                    "flexMs" to flexMs,
                    "requiresNetwork" to requiresNetwork,
                    "requiresCharging" to requiresCharging,
                    "requiresDeviceIdle" to requiresDeviceIdle,
                    "runBudgetMs" to runBudgetMs,
                    "foregroundStopTimeoutMs" to foregroundStopTimeoutMs,
                    "callbackHandle" to callbackHandle,
                )
            )
            .toString()

    companion object {
        fun decode(
            json: JSONObject,
            callbackHandle: Long = json.getLong("callbackHandle"),
        ): TaskConfiguration {
            fun duration(key: String): Long? =
                if (json.isNull(key)) null
                else json.getLong(key).also { require(it >= 0) { "$key must not be negative" } }

            val identifier = json.getString("identifier")
            val kind = json.getString("kind")
            val frequency = json.getLong("frequencyMs")
            val flex = duration("flexMs")
            require(identifier.isNotBlank()) { "Task identifier must not be blank" }
            require(kind == "refresh" || kind == "processing") {
                "Task kind must be refresh or processing"
            }
            require(frequency >= 15 * 60 * 1000L) { "Frequency must be at least 15 minutes" }
            require(flex == null || flex in 5 * 60 * 1000L..frequency) {
                "Flex must be at least 5 minutes and no greater than frequency"
            }
            require(callbackHandle != 0L) { "A retained background dispatcher is required" }
            return TaskConfiguration(
                identifier,
                kind,
                frequency,
                duration("initialDelayMs") ?: 0,
                flex,
                json.optBoolean("requiresNetwork"),
                json.optBoolean("requiresCharging"),
                json.optBoolean("requiresDeviceIdle"),
                duration("runBudgetMs"),
                duration("foregroundStopTimeoutMs"),
                callbackHandle,
            )
        }
    }
}

File: mobile/apps/photos/plugins/ente_background_manager/ios/Classes/TaskConfiguration.swift
import Foundation

struct TaskConfiguration: Codable, Equatable {
  let identifier: String
  let kind: String
  let frequencyMs: Int64
  let initialDelayMs: Int64
  let flexMs: Int64?
  let requiresNetwork: Bool
  let requiresCharging: Bool
  let requiresDeviceIdle: Bool
  let runBudgetMs: Int64?
  let foregroundStopTimeoutMs: Int64?
  let callbackHandle: Int64

  init(_ data: [String: Any], callbackHandle: Int64) throws {
    guard let identifier = data["identifier"] as? String, !identifier.isEmpty,
      let kind = data["kind"] as? String, ["refresh", "processing"].contains(kind),
      let frequency = data["frequencyMs"] as? NSNumber, frequency.int64Value > 0,
      callbackHandle != 0
    else {
      throw ConfigurationError.invalidTask
    }
    self.identifier = identifier
    self.kind = kind
    self.frequencyMs = frequency.int64Value
    self.initialDelayMs = (data["initialDelayMs"] as? NSNumber)?.int64Value ?? 0
    self.flexMs = (data["flexMs"] as? NSNumber)?.int64Value
    self.requiresNetwork = data["requiresNetwork"] as? Bool ?? false
    self.requiresCharging = data["requiresCharging"] as? Bool ?? false
    self.requiresDeviceIdle = data["requiresDeviceIdle"] as? Bool ?? false
    self.runBudgetMs = (data["runBudgetMs"] as? NSNumber)?.int64Value
    self.foregroundStopTimeoutMs = (data["foregroundStopTimeoutMs"] as? NSNumber)?.int64Value
    self.callbackHandle = callbackHandle
    guard
      [initialDelayMs, flexMs, runBudgetMs, foregroundStopTimeoutMs]
        .compactMap({ $0 }).allSatisfy({ $0 >= 0 })
    else {
      throw ConfigurationError.invalidDuration
    }
    guard flexMs == nil, !requiresDeviceIdle,
      kind == "processing" || (!requiresNetwork && !requiresCharging)
    else {
      throw ConfigurationError.unsupportedConstraint
    }
  }

  enum ConfigurationError: Error {
    case invalidTask
    case invalidDuration
    case unsupportedConstraint
    case unregisteredIdentifier
    case duplicateIdentifier
  }
}

struct StoredConfiguration: Codable {
  var enabled: Bool
  var tasks: [TaskConfiguration]

  static func load() -> StoredConfiguration {
    guard let data = UserDefaults.standard.data(forKey: "ente_background_manager"),
      let value = try? JSONDecoder().decode(StoredConfiguration.self, from: data)
    else {
      return StoredConfiguration(enabled: false, tasks: [])
    }
    return value
  }

  func save() throws {
    UserDefaults.standard.set(try JSONEncoder().encode(self), forKey: "ente_background_manager")
  }
}
